### ISSUE TITLE
Badly behaved comparison and/or "if" operator in compiled binary

### ISSUE THREAD

**Issue by Chant** [justinethier/cyclone#248]
I expect the following script to output the following sequence of four characters: 1, newline, 1, newline.
```
(import (scheme base) (scheme write))

(define (greater-of x y) (if (> x y) x y))

(write (greater-of 0 1))
(newline)
(write (greater-of 1 0))
(newline)
```

It behaves as expected when interpreted via `$ icyc < greater-of.scm` (accounting for extra output from prompts and such).

It does not behave as expected when compiled and then executed via `$ cyclone greater-of.scm && ./greater-of`. Its output is 1, newline, 0, newline.

An inverse script `lesser-of` where the `<` operator is used in the same way fails with precisely the same output. Rephrasing the definition of `greater-of` using a lambda instead of the define procedure special form does not change the output.

My `cyclone` was compiled by checking out tag `v0.7.2` of `cyclone-bootstrap` and executing `make` followed by `sudo make install`. No error messages were produced in these operations. Retrying the bootstrap in full does not change the results I see.

I'm really hoping I'm missing some obvious flaw in my expectations!

**Comment by Chant** [comment]
Ah, and:
```
$ make test
make unit-tests
make[1]: Entering directory '/home/chant/src/cyclone-bootstrap'
./cyclone unit-tests.scm && ./unit-tests
(109 tests passed with no errors)
make[1]: Leaving directory '/home/chant/src/cyclone-bootstrap'
```

**Comment by justinethier** [comment]
@Chant - Thanks for the report! It looks like what is happening is the compiler is introducing multiple copies of the same lambda and variables during beta expansion, which causes the optimizer to pick up the wrong value when optimizing-out calls to `greater-of`. Will need to dig deeper to find a solution.

A quick short term work-around is to disable optimizations with `-O0`.

**Comment by justinethier** [comment]
This is fixed now on master and will be included in the next release (0.7.3). Thanks again for the report, @Chant !!

**Comment by Chant** [comment]
@justinethier, thanks in turn for the shockingly quick resolution! The workaround you advised has been effective for me in the meantime.
